src/Microsoft.ML.TimeSeries/RootCauseAnalyzer.cs

```diff
@@ -0,0 +1,764 @@
+                if (slice.TimeStamp.Equals(src.AnomalyTimestamp))
+                {
+                    points = slice.Points;
+                }
+            }
+
+            return points;
+        }
+
+        protected DimensionInfo SeperateDimension(Dictionary<string, Object> dimensions, Object aggSymbol)
+        {
+            DimensionInfo info = DimensionInfo.CreateDefaultInstance();
+            foreach (KeyValuePair<string, Object> entry in dimensions)
+            {
+                string key = entry.Key;
+                if (aggSymbol.Equals(entry.Value))
+                {
+                    info.AggDims.Add(key);
+                }
+                else
+                {
+                    info.DetailDims.Add(key);
+                }
+            }
+
+            return info;
+        }
+
+        protected Tuple<PointTree, PointTree, Dictionary<string, Point>> GetPointsInfo(RootCauseLocalizationInput src, DimensionInfo dimensionInfo)
+        {
+            PointTree pointTree = PointTree.CreateDefaultInstance();
+            PointTree anomalyTree = PointTree.CreateDefaultInstance();
+            Dictionary<string, Point> dimPointMapping = new Dictionary<string, Point>();
+
+            List<Point> totalPoints = GetTotalPointsForAnomalyTimestamp(src);
+            Dictionary<string, Object> subDim = GetSubDim(src.AnomalyDimension, dimensionInfo.DetailDims);
+
+            foreach (Point point in totalPoints)
+            {
+                if (ContainsAll(point.Dimension, subDim))
+                {
+                    if (!dimPointMapping.ContainsKey(GetDicCode(point.Dimension)))
+                    {
+                        dimPointMapping.Add(GetDicCode(point.Dimension), point);
+                        bool isValidPoint = point.IsAnomaly == true;
+                        if (ContainsAll(point.Dimension, subDim))
+                        {
+                            BuildTree(pointTree, dimensionInfo.AggDims, point, src.AggSymbol);
+
+                            if (isValidPoint)
+                            {
+                                BuildTree(anomalyTree, dimensionInfo.AggDims, point, src.AggSymbol);
+                            }
+                        }
+                    }
+                }
+            }
+
+            return new Tuple<PointTree, PointTree, Dictionary<string, Point>>(pointTree, anomalyTree, dimPointMapping);
+        }
+
+        protected Dictionary<string, Object> GetSubDim(Dictionary<string, Object> dimension, List<string> keyList)
+        {
+            Dictionary<string, Object> subDim = new Dictionary<string, Object>();
+            foreach (string dim in keyList)
+            {
+                subDim.Add(dim, dimension[dim]);
+            }
+            return subDim;
+        }
+
+        protected List<RootCauseItem> LocalizeRootCauseByDimension(PointTree anomalyTree, PointTree pointTree, Dictionary<string, Object> anomalyDimension, List<string> aggDims)
```

Comment:
suggestion: Please consider something along the lines of the following:
```
return new Dictionary<string, object>(keyList.Select(dim => new KeyValuePair<string, object>(dim, dimension[dim])));

```

Reply:
> suggestion: Please consider something along the lines of the following for the GetSubDim function.
> 
> ```
> return new Dictionary<string, object>(keyList.Select(dim => new KeyValuePair<string, object>(dim, dimension[dim])));
> ```

updated